libsignal-protocol/src/crypto.rs

```diff
@@ -1,23 +1,266 @@
+    fn hmac_sha256_update(&mut self, data: &[u8]) -> Result<(), InternalError>;
+    fn hmac_sha256_final(&mut self) -> Result<Vec<u8>, InternalError>;
+    fn hmac_sha256_cleanup(&mut self) {}
+
+    fn sha512_digest_init(&mut self) -> Result<(), InternalError>;
+    fn sha512_digest_update(
+        &mut self,
+        data: &[u8],
+    ) -> Result<(), InternalError>;
+    fn sha512_digest_final(&mut self) -> Result<Vec<u8>, InternalError>;
+    fn sha512_digest_cleanup(&mut self) {}
+
+    fn encrypt(
+        &self,
+        cipher: SignalCipherType,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError>;
+    fn decrypt(
+        &self,
+        cipher: SignalCipherType,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError>;
 }
 
+#[cfg(not(target_os = "linux"))]
 #[derive(Debug, Copy, Clone, PartialEq)]
 pub struct DefaultCrypto;
 
+#[cfg(target_os = "linux")]
+#[derive(Clone)]
+pub struct DefaultCrypto {
+    hmac_ctx: Option<openssl::hash::Hasher>,
+    sha512_ctx: Option<openssl::hash::Hasher>,
+}
+
+#[cfg(not(target_os = "linux"))]
 impl Crypto for DefaultCrypto {
-    fn fill_random(&self, buffer: &mut [u8]) {
+    fn fill_random(&self, buffer: &mut [u8]) -> Result<(), InternalError> {
+        use rand::RngCore;
         let mut rng = rand::thread_rng();
         rng.fill_bytes(buffer);
+        Ok(())
+    }
+
+    fn hmac_sha256_init(&mut self, key: &[u8]) -> Result<(), InternalError> {
+        unimplemented!()
+    }
+
+    fn hmac_sha256_update(&mut self, data: &[u8]) -> Result<(), InternalError> {
+        unimplemented!()
+    }
+
+    fn hmac_sha256_final(&mut self) -> Result<Vec<u8>, InternalError> {
+        unimplemented!()
+    }
+
+    fn sha512_digest_init(&mut self) -> Result<(), InternalError> {
+        unimplemented!()
+    }
+
+    fn sha512_digest_update(
+        &mut self,
+        data: &[u8],
+    ) -> Result<(), InternalError> {
+        unimplemented!()
+    }
+
+    fn sha512_digest_final(&mut self) -> Result<Vec<u8>, InternalError> {
+        unimplemented!()
+    }
+
+    fn encrypt(
+        &self,
+        cipher: SignalCipherType,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError> {
+        unimplemented!()
+    }
+
+    fn decrypt(
+        &self,
+        cipher: SignalCipherType,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError> {
+        unimplemented!()
+    }
+}
+
+#[cfg(not(target_os = "linux"))]
+impl Default for DefaultCrypto {
+    fn default() -> Self { Self }
+}
+
+#[cfg(target_os = "linux")]
+impl Default for DefaultCrypto {
+    fn default() -> Self {
+        Self {
+            hmac_ctx: None,
+            sha512_ctx: None,
+        }
+    }
+}
+#[cfg(target_os = "linux")]
+impl DefaultCrypto {
+    pub fn crypter(
```

Comment:
does this need to be `pub`? I don't think it'd be of use to anyone outside this file.

Reply:
Yeah, you alright, it should be `pub(in crypto)` for even module privacy 😌

Reply:
if that's the case... why not just drop the `pub` altogether?

Reply:
yeah, makes more sense here :+1: